src/main/java/org/veupathdb/service/eda/ss/model/reducer/DataFlowTreeFactory.java

```diff
@@ -33,8 +34,8 @@ public TreeNode<DataFlowNodeContents> create(TreeNode<Entity> prunedEntityTree,
                                                Study study) {
     final TreeNode<Entity> outputNode = prunedEntityTree.findFirst(entity -> entity.equals(outputEntity));
     // In lieu of a pointer up the tree, we use this function to traverse from the original root to a node's parent.
-    final Function<TreeNode<Entity>, TreeNode<Entity>> parentMapper = child ->
-        prunedEntityTree.findFirst(candidate -> candidate.getChildNodes().contains(child), null);
+    final Function<TreeNode<Entity>, Optional<TreeNode<Entity>>> parentMapper = child ->
```

Comment:
a comment saying why we need optionals here might help

Reply:
Added clarification here